### PR TITLE
[DTensor][1/N] add DTensor RNG state APIs

### PR DESCRIPTION
**Pull request by XilunWu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98577
* #98199
* __->__ #98198

### CODE REVIEW

File: torch/distributed/_tensor/ops/pointwise_ops.py [hunk 1]
```diff
@@ -382,19 +382,9 @@ def _register_non_deterministic_op(op):
...
-            if isinstance(placement, (Replicate, _Partial)):
-                replicate_or_partial = True
-                break
-
-        if replicate_or_partial:
-            return OutputSharding(
-                None, failed_reason=f"{op} with replication is not supported yet!"
-            )
+        # TODO: we will support native_dropout in future
+        if op == aten.native_dropout.default:
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/pointwise_ops.py:386]: why we remove it as a whole? We used to support the dropout with shard placement

File: torch/distributed/_tensor/device_mesh.py [hunk 2]
```diff
@@ -152,6 +153,11 @@ def _get_or_create_default_group(self):
                 # automatically set the current cuda device base on num of gpu devices available in each host
                 # NOTE: This device selection would only work for homogenous hardware.
                 torch.cuda.set_device(get_rank() % torch.cuda.device_count())
+            # to perform DTensor random ops, we need to ensure all ranks in mesh is initialized with the same
+            # random seed. The seed to use will be the current seed on rank 0.
+            object_list = [torch.cuda.initial_seed()]
+            broadcast_object_list(object_list)
+            torch.cuda.manual_seed(int(object_list[0]))
```
Comment (wanchaol) [inline torch/distributed/_tensor/device_mesh.py:160]: I am not sure if we should by default manual_seed here in the device mesh constructor(doing it in `manual_seed` call is ok), we should probably first assign the initial seed as an attribute of device_mesh. And we need to think about how to use it.

File: torch/distributed/_tensor/ops/pointwise_ops.py (hunk 1 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/pointwise_ops.py:386]: You're right. Even after using the new RNG approach, `dropout` with `_Partial` is still not supported so the current logic should be kept.

File: torch/distributed/_tensor/device_mesh.py (hunk 2 above)
Comment (XilunWu) [reply torch/distributed/_tensor/device_mesh.py:160]: sounds good.

File: torch/distributed/_tensor/random.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        `manual_seed` will not set its GPU device's generator seed.
+        Current implementation only supports a GPU device mesh.
+    """
+    assert (
+        device_mesh is not None
+    ), f"expect a DeviceMesh parameter but {device_mesh} was passed in."
+
+    import torch.distributed as dist
+
+    # broadcast the seed from rank 0 over the default PG
```
Comment (wanchaol) [inline torch/distributed/_tensor/random.py:83]: nit: allgather the seed

File: torch/distributed/_tensor/ops/random_ops.py [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+            replicate_or_partial = True
+            break
+
+    if replicate_or_partial:
+        return OutputSharding(
+            None,
+            failed_reason="aten.native_dropout.default with replication is not supported yet!",
+        )
+    else:
+        return OutputSharding(self_spec)
```
Comment (wanchaol) [inline torch/distributed/_tensor/ops/random_ops.py:54]: hmmm I thought `native_dropout` returns two tensors instead of one?

File: test/distributed/_tensor/test_random_ops.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+        for shard_num in range(self.world_size):
+            if self.rank == shard_num:
+                self.assertEqual(local_tensor[:, shard_num], local_tensor[:, self.rank])
+            else:
+                self.assertNotEqual(
+                    local_tensor[:, shard_num], local_tensor[:, self.rank]
+                )
+
+        # TODO: support dropout
+        with self.assertRaisesRegex(RuntimeError, "supported"):
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:85]: nit: this should be a different test case

File: test/distributed/_tensor/test_random_ops.py [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+        # postprocess rng offset
+        _set_offset(offset + global_size, device_mesh)
+        check_rng_state(1234, global_size, device_mesh)
+
+        dtensor = dtensor.redistribute(device_mesh, [Replicate()])
+        local_tensor = dtensor.to_local()
+
+        for shard_num in range(self.world_size):
+            if self.rank == shard_num:
+                self.assertEqual(local_tensor[:, shard_num], local_tensor[:, self.rank])
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:78]: hmmmm what are we comparing here, if `self.rank == shard_num, this is always equal...

File: test/distributed/_tensor/test_random_ops.py [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+            else:
+                self.assertNotEqual(
+                    local_tensor[:, shard_num], local_tensor[:, self.rank]
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:81]: is this trying to compare that different shard have different values?

**Review by wanchaol — COMMENTED** [review]
Looks much better! I have some questions about the test case of uniform, comments inlined.

File: torch/distributed/_tensor/device_mesh.py [hunk 8]
```diff
@@ -152,6 +153,12 @@ def _get_or_create_default_group(self):
                 # automatically set the current cuda device base on num of gpu devices available in each host
                 # NOTE: This device selection would only work for homogenous hardware.
                 torch.cuda.set_device(get_rank() % torch.cuda.device_count())
+            # to perform DTensor random ops, we need to ensure all ranks in mesh is initialized with the same
+            # random seed. The seed to use will be the current seed on rank 0. We store this seed as an attribute
+            # of device mesh for future use.
+            object_list = [torch.cuda.initial_seed()]
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/device_mesh.py:159]: The reason that I use `initial_seed()` is, this API returns an `int` value (i.e. `seed`) but `get_rng_state()` returns a `ByteTensor` whose size is around 400 bytes (`seed`, `offset`, and padding). 

Right now, we ask users to call DTensor's `manual_seed()` after initializing a device mesh. And we likely want to use the seed and initialize it somewhere in future. In both cases, `offset` will be set to 0 by `manual_seed()` so there is no need to store `offset`.

File: torch/distributed/_tensor/random.py [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+    Returns:
+        A :class:`Tensor` object that contains the random number generator state.
+
+    .. warning::
+        Current implementation only supports a GPU device mesh.
+        If ``device_mesh`` is a sub-mesh and the calling rank is not a part of it,
+        `get_rng_state` still returns its GPU device's generator state.
+    """
+    assert (
+        device_mesh is not None
```
[the remark this answers is not in the document]
Comment (XilunWu) [reply torch/distributed/_tensor/random.py:48]: for function
```
def f(x: int) -> int:
    return x
```
the caller can still pass `None` as argument. e.g. `f(None)`.

Using `isinstance` can also error out the case.

File: torch/distributed/_tensor/random.py (hunk 3 above)
Comment (XilunWu) [reply torch/distributed/_tensor/random.py:83]: thx for catching it!

File: test/distributed/_tensor/test_random_ops.py (hunk 6 above)
Comment (XilunWu) [reply test/distributed/_tensor/test_random_ops.py:78]: right... change the loop body to:
```
if rank != shard_num:
    self.assertNotEqual(local_tensor[:, shard_num], local_tensor[:, rank])
```

File: test/distributed/_tensor/test_random_ops.py (hunk 7 above)
Comment (XilunWu) [reply test/distributed/_tensor/test_random_ops.py:81]: correct.

File: torch/distributed/_tensor/ops/random_ops.py (hunk 4 above)
Comment (XilunWu) [reply torch/distributed/_tensor/ops/random_ops.py:54]: I want to enable `native_dropout` in another PR so I kept this part untouched except for enabling `uniform` and `normal`.

File: torch/distributed/_tensor/device_mesh.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (wanchaol) [reply torch/distributed/_tensor/device_mesh.py:159]: I see. sounds good, given that we don't really know how to use this `_seed` yet, shall we remove the broadcast initialization logic first and add it later once we know how it should be used?

We are using the `manual_seed` API anyways in the tests. I am a little bit worried that this additional broadcast object might not be friendly to internal systems.

File: test/distributed/_tensor/test_random_ops.py [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+        # postprocess rng offset
+        _set_offset(offset + global_size, device_mesh)
+        check_rng_state(1234, global_size, device_mesh)
+
+        # allgather the local tensors
+        local_tensor = dtensor.to_local()
+        local_tensor_list = [
+            torch.empty_like(local_tensor) for i in range(self.world_size)
+        ]
+        device_mesh.all_gather(local_tensor_list, local_tensor)
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:78]: does it work? can you rebase to the lastest mater? I think `mesh.allgather` changes its API to return a single bigger tensor instead of tensor list. So I think this test should crash. 

Can you just call `dtensor.redistribute([Replicate()]).to_local()` instead of manually call allgather here? we should try to use redistribute conversion instead of individual collective imo.

File: test/distributed/_tensor/test_random_ops.py [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+        with self.assertRaisesRegex(RuntimeError, "supported"):
+            dropout = torch.nn.Dropout(p=0.2)
+            dtensor = dropout(dtensor)
+
+            # allgather the local tensors
+            local_tensor = dtensor.to_local()
+            local_tensor_list = [
+                torch.empty_like(local_tensor) for i in range(self.world_size)
+            ]
+            device_mesh.all_gather(local_tensor_list, local_tensor)
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:115]: For this case, it looks like you already call redistribute to replicate eariler, I assume you want to test the dropout on replicate case, probably need to use dist.all_gather instead of device_mesh.allgather

File: test/distributed/_tensor/test_random_ops.py [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+        local_tensor = dtensor.to_local()
+        local_tensor_list = [
+            torch.empty_like(local_tensor) for i in range(self.world_size)
+        ]
+        device_mesh.all_gather(local_tensor_list, local_tensor)
+
+        # compare with local tensors from other ranks
+        for other_rank in range(self.world_size):
+            if self.rank != other_rank:
+                # other rank should have an identical local tensor
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:83]: I think it should have different results instead of "identical"?

File: test/distributed/_tensor/test_random_ops.py [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+        # initialize rng state
+        manual_seed(1234, device_mesh)
+        _tensor = torch.empty(*size, device="cuda")
+        dtensor = DTensor.from_local(_tensor, device_mesh, [Shard(1)])
+
+        # a random op call shifts the offset
+        dtensor.uniform_(0, 1)
+
+        # the dtensor is now replicate on all ranks
+        dtensor = dtensor.redistribute(device_mesh, [Replicate()])
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:103]: do you want to do something with this redistribute, or is it being used in the later part of test?

**Review by wanchaol — COMMENTED** [review]
I think you need to rebase and fix the tests (please make sure it succeed locally), as `device_mesh.allgather` won't take list of tensors.

File: test/distributed/_tensor/test_random_ops.py (hunk 10 above)
Comment (XilunWu) [reply test/distributed/_tensor/test_random_ops.py]: replaced `allgather` call with a `redistribute([Replicate()])`.

File: test/distributed/_tensor/test_random_ops.py (hunk 11 above)
Comment (XilunWu) [reply test/distributed/_tensor/test_random_ops.py]: replaced `allgather` with ```
DTensor.from_local(dtensor.to_local(), mesh, [Shard(0)])
redistribute([Replicate()])
```

File: test/distributed/_tensor/test_random_ops.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+        # allgather the local tensors
+        dtensor = dtensor.redistribute(device_mesh, [Replicate()])
+        local_tensor = dtensor.to_local()
+
+        # compare with local tensors from other ranks
+        for other_rank in range(self.world_size):
+            if self.rank != other_rank:
+                # other rank should have a different local tensor
+                self.assertNotEqual(
+                    local_tensor[:, self.rank], local_tensor[:, other_rank]
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:89]: hmmm is this really correct? if self.rank != other_rank, `local_tensor[:, self.rank]` and `local_tensor[:, other_rank]` would have different shape, so they are definitely not equal. I guess you are not testing what you want, you probably want to do things like `local_tensor_list = local_tensor.tensor_split(4)`, then compare the list of tensors instead.

File: test/distributed/_tensor/test_random_ops.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+            local_tensor = dtensor.to_local()
+
+            # compare with local tensors from other ranks
+            self_slice = slice(4 * self.rank, 4 * self.rank + 4)
+            for other_rank in range(self.world_size):
+                if self.rank != other_rank:
+                    # other rank should have an identical local tensor
+                    other_slice = slice(4 * other_rank, 4 * other_rank + 4)
+                    self.assertEqual(
+                        local_tensor[self_slice, :], local_tensor[other_slice, :]
```
Comment (wanchaol) [inline test/distributed/_tensor/test_random_ops.py:126]: looks like this might be correct (not exactly sure), but please align the comparison using the same approach so that it's easy to read. (you can choose to use this one or the one I mentioned in the above comment)

File: torch/distributed/_tensor/random.py [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+        None
+
+    .. warning::
+        Current implementation only supports a GPU device mesh.
+        If ``device_mesh`` is a sub-mesh and the calling rank is not a part of it,
+        `set_rng_state` will not set its GPU device's generator state.
+    """
+    assert isinstance(
+        device_mesh, DeviceMesh
+    ), f"expect a DeviceMesh but {device_mesh} was passed in."
```
Comment (wanchaol) [inline torch/distributed/_tensor/random.py:26]:
```suggestion
    ), f"expect a DeviceMesh but {type(device_mesh)} was passed in."
```

File: torch/distributed/_tensor/random.py [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+        A :class:`torch.ByteTensor` object that contains the random number generator state.
+
+    .. warning::
+        Current implementation only supports a GPU device mesh.
+        If ``device_mesh`` is a sub-mesh and the calling rank is not a part of it,
+        `get_rng_state` still returns its GPU device's generator state.
+    """
+    assert isinstance(
+        device_mesh, DeviceMesh
+    ), f"expect a DeviceMesh but {device_mesh} was passed in."
```
Comment (wanchaol) [inline torch/distributed/_tensor/random.py:55]: same suggestion as above

File: torch/distributed/_tensor/random.py [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        When calling this function, :func:`manual_seed` must be called from all ranks of the
+        default `ProcessGroup` even if some ranks may not be a part of the `device_mesh`,
+        with the same `seed` value.
+        If ``device_mesh`` is a sub-mesh and the calling rank is not a part of it,
+        `manual_seed` will not set its GPU device's generator seed.
+        Current implementation only supports a GPU device mesh.
+    """
+    assert isinstance(
+        device_mesh, DeviceMesh
+    ), f"expect a DeviceMesh but {device_mesh} was passed in."
```
Comment (wanchaol) [inline torch/distributed/_tensor/random.py:85]: same suggestion as above.

**Review by wanchaol — APPROVED** [review]
lgtm, I have some additional suggestions with regards to tests, see inlined comments.

File: test/distributed/_tensor/test_random_ops.py (hunk 14 above)
Comment (XilunWu) [reply test/distributed/_tensor/test_random_ops.py:89]: Let me quickly double check it

Comment (XilunWu) [reply test/distributed/_tensor/test_random_ops.py:89]: resolved offline
